doc/user-guide.md

```diff
@@ -430,54 +430,90 @@ deleted until you remove the finalizer (ie, after your cleanup logic has success
 
 ```Go
+
+const memcachedFinalizer = "finalizer.cache.example.com"
+
 func (r *ReconcileMemcached) Reconcile(request reconcile.Request) (reconcile.Result, error) {
  	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
  	reqLogger.Info("Reconciling Memcached")
  
  	// Fetch the Memcached instance
  	memcached := &cachev1alpha1.Memcached{}
  	err := r.client.Get(context.TODO(), request.NamespacedName, memcached)
+
  	...
+
  	// Check if the APP CR was marked to be deleted
  	isMemcachedMarkedToBeDeleted := memcached.GetDeletionTimestamp() != nil
  	if isMemcachedMarkedToBeDeleted {
- 		// TODO(user): Add the cleanup steps that the operator needs to do before the CR can be deleted
- 		// Update finalizer to allow delete CR
- 		memcached.SetFinalizers(nil)
- 
- 		// Update CR
- 		err := r.client.Update(context.TODO(), memcached)
- 		if err != nil {
- 			return reconcile.Result{}, err
- 		}
+		if contains(memcached.GetFinalizers(), memcachedFinalizer) {
+			// Run finalization logic for memcachedFinalizer. If the
+			// finalization logic fails, don't remove the finalizer so
+			// that we can retry during the next reconciliation.
+			if err := r.finalizeMemcached(); err != nil {
+			     return reconcile.Result{}, err
+			}
+
+			// Remove memcachedFinalizer. Once all finalizers have been
+			// removed, the object will be deleted.
+			memcached.SetFinalizers(remove(memcached.GetFinalizers(), memcachedFinalizer))
+			err := r.client.Update(context.TODO(), memcached)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
  		return reconcile.Result{}, nil
  	}
  	
  	// Add finalizer for this CR
- 	if err := r.addFinalizer(reqLogger, instance); err != nil {
- 		return reconcile.Result{}, err
- 	}
+	if !contains(memcached.GetFinalizers(), memcachedFinalizer) {
+		if err := r.addFinalizer(reqLogger, instance); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
  	...
  
  	return reconcile.Result{}, nil
  }
  
-//addFinalizer will add this attribute to the Memcached CR
+func (r *ReconcileMemcached) finalizeMemcached(reqLogger logr.Logger, m *cachev1alpha1.Memcached) error {
+	// TODO(user): Add the cleanup steps that the operator
+	// needs to do before the CR can be deleted
+	reqLogger.Info("Successfully finalized memcached")
+	return nil
+}
+
 func (r *ReconcileMemcached) addFinalizer(reqLogger logr.Logger, m *cachev1alpha1.Memcached) error {
-    if len(m.GetFinalizers()) < 1 && m.GetDeletionTimestamp() == nil {
-        reqLogger.Info("Adding Finalizer for the Memcached")
-        m.SetFinalizers([]string{"finalizer.cache.example.com"})
-    
-        // Update CR
-        err := r.client.Update(context.TODO(), m)
-        if err != nil {
-            reqLogger.Error(err, "Failed to update Memcached with finalizer")
-            return err
-        }
-    }
-    return nil
+	reqLogger.Info("Adding Finalizer for the Memcached")
```

Comment:
need to specify correct ref for logging package. `logr.Logger` is not recognized.
`var log = logf.Log.WithName("controller_snatip")` --> need to give ref based on this one

Reply:
Unless I'm misunderstanding something, I think this is fine because the first line of the scaffolded `Reconcile` method sets `reqLogger := log.WithValues(...)`.

Reply:
actually, there is no need to add logger param in this function. its nice to have though.